JupyterLab 3.0 release survey (#107)
jupyterlab/team-compass: docs/surveys/2020-jupyterlab-survey.md

# JupyterLab

 - Even if you don't use Jupyter, you can still take this survey. Just indicate this fact in the first question and carry on as best as you can.

>  - Thank you - your participation guides Jupyter's roadmap toward your real-life use cases.

Reviewer: Can we add a couple of sentences about what we plan to do with the survey data, and how privacy will be managed, and who will see the survey data?

Reviewer 2: @jasongrout is this along the lines of what you were going for?

"The aggregate survey data itself will be openly shared with the Jupyter community when polling closes in mid-December. If you opt to provide your email for a user interview, it not be used for Jupyter's promotional purposes and it will not be shared with a 3rd party."
^ Figured we will run it for a month depending on how many responses we get?

"Thank you. Your participation guides Jupyter's roadmap toward your real-life use cases by quantifiably helping us prioritize the functionality that is important to our userbase."

>  - So that you know what to expect, it's comprised of 20 questions (#7 is the only big one) spread across these sections:

Reviewer 2: Would prefer to set expectations by directly confronting the issue with transparency rather than to blindside them. This approach has worked for me in many customer-facing scenarios. It also builds trust. I rewording it to be truer: " As a fair heads up, Question 7  is the biggest one, but it provides critical information."

>  - So that you know what to expect, it's comprised of 20 questions (#7 is the only big one) spread across these sections:

Reviewer 3: I like your suggested change 👍 Thanks @LayneSadler :sunny:

    - Usage patterns.
    - Data
    - Visualization.
    - Scale.
    - Collaboration.

### Usage Patterns

1. How frequently do you use Jupyter?

>     - Daily: 4 or more hours per day.

Reviewer 2: ☑️ change integrated.

    - Daily: 3 or less hours per day. 
    - Weekly.
    - Monthly.
    - I no longer use Jupyter.
    - I have never used Jupyter.

2. How long have you been using Jupyter?
    - 2+ years.
    - 1-2 years.
    - 6-12 months.
    - Less than 6 months (welcome =]).
    - I don't use Jupyter.

> 3. What languages do you use in Jupyter? (pick up to 4)

Reviewer 2: ☑️  change integrated.

> 3. What languages do you use in Jupyter? (pick up to 4)

Reviewer 3: Oh my, how did I forget "Julia"? Please add Julia here.

    - C (and derivatives)
    - Groovy
    - Java
    - JavaScript
    - NodeJS
    - Perl
    - PHP
    - Python
    - R
    - Ruby
    - Rust
    - Scala
    - Spark SQL
    - SQL
    - TypeScript
    - ❗I wrap/ use bindings for other languages.
    - ❗My preferred language is not supported in Jupyter.
    - Other (please specify)

4. What are your primary job roles when you are using Jupyter? (pick up to 2)
    - Backend engineer.
    - Business analyst.
    - Data engineer.
    - Data scientist.
    - Database Admin (DBA).
    - DevOps.
    - Financial modeler/ analyst.
    - Front end/ web development.
    - Infrastructure engineer/ cloud architect.
    - Scientist/ researcher.
    - Student.
    - Sysadmin.
    - Teacher/ lecturer.
    - Tutor/ teaching assistant.
    - Other (please specify)

5. What are your go-to tools for performing data science, scientific computing, 
and machine learning on your laptop/ desktop (non-cloud) for data science? (pick up to 3)

>     - Spyder.

Reviewer 2: ☑️  change integrated. Alphabetized where ever  it made sense.

    - PyCharm.
    - nteract.
    - Atom.
    - Emacs.
    - Zeppelin.
    - iPython.
    - VS Code.
    - JupyterLab
    - RStudio.
    - Jupyter Notebook (classic)
    - Vim.
    - Sublime Text.

>     - Other (please specify)

Reviewer 4: How are the options presented to the user:
- Randomly with `other` at the end
- Alphabetically (does not seem the current case)
- Should it have some specific order?

Reviewer 2: Yeah, random seems almost too randomly for the sake of the time it takes to complete. Alphabetize it is.

Reviewer: Totally minor nitpick: some options have periods, some don't

Reviewer 2: ☑️  changes integrated. Alphabetized wherever it made sense. Used periods everywhere; it reflects syntax/ cody tidiness.

6. How do you run and/ or access Jupyter? (pick up to 4)
    - Run directly on local machine (laptop, desktop).
    - Through a Python virtual environment (e.g. conda, venv)
    - Through Docker.
    - HPC or on-premise server.
    - Cloud server (e.g. EC2).
    - JupyterHub.
    - BinderHub / MyBinder.
    - Cloud service - AWS: EMR, SageMaker.
    - Cloud service - Azure: Notebooks, ML Studio.
    - Cloud service - Databricks
    - Cloud service - Google: AI Platform, Dataproc.
    - Cloud service - IBM: Watson.
    - Google Colab.

>     - Through a 3rd party SaaS app (e.g. Observable, CoCalc)

Reviewer: I think CoCalc should be listed in the Cloud service section as its own item. I'm not sure why it is listed as a "3rd party", but others are listed individually. I think "SaaS" is probably too technical of an abbreviation to use for this survey - I would suggest either deleting this item (letting them fill out "Other").

ObservableHQ doesn't offer Jupyter access, does it? Is that the Observable mentioned here?

Reviewer 3: Agree CoCalc deserves its own item

Reviewer 2: Will make these adjustments and drop SaaS. I think I had PaaS (clouds) and SaaS, but brought out the PaaS so it makes sense to break out the SaaS too. Will drop Observable.

Reviewer 2: ☑️ changes integrated.

    - ❓Don’t know how, I just go to a URL.
    - Other (please specify)

> 7. What tasks do you need to perform and what tools do you use to accomplish them?

Reviewer 3: This is a pretty long question. You may lose responses with this questions as folks may find the question burdensome.

Reviewer 2: Agree, we definitely will! But this is the data we need. Confident we can still get a statistically significant sample size. This matrix approach came from the community.

Reviewer 3: @LayneSadler I just saw the rendered version which is less scary than the way it is displayed in markdown. Go ahead and mark my comment resolved :sunny:

    - Writing a software package.
    - Visualize data in charts, plots, or dashboards.
    - Writing software documentation.
    - Building a machine learning or statistical model.
    - Cleaning and preparing data.

>     - Run pipelines, workflows, or ETL jobs.

Reviewer 4: Maybe expand on what ETL means (just to be as clear as we can?)

Reviewer 2: ☑️ changes integrated; spelled out extract, transform, load

    - Developing extensions/ plugins to solve my problems.
    - Finding extensions/ plugins to solve my problems.

>     - Creating content (blogs).

Reviewer 4: Adding "examples" in parens, sounds good, but having only one example feels a bit lonely?

Reviewer: Perhaps also "books", "education materials"

Reviewer 2: ☑️ changes integrated; included books and education materials

>     - Documenting research (reports, scientific papers)

Reviewer 4: Adding `e.g.` in parens  or similar would make thise clearer?

Reviewer 2: ☑️ changes integrated. Used 'e.g.' everywhere; it reflects syntax/ cody tidiness.

        - Does not apply.
        - Never meets.
        - Rarely meets.
        - Sometimes meets.
        - Often exceeds.
        - Always exceeds.
    - To what degree do alternative tools meet your expectations for this?
        - Does not apply.
        - Never meets.
        - Rarely meets.
        - Sometimes meets.
        - Often exceeds.
        - Always exceeds.

### Data

8. What data sources are you primarily working with in your role? (pick up to 3)
    - My local file system (e.g. files and folder on local machine).
    - File system (e.g. HPC, EBS/EFS, JupyterHub volumes).
    - Cloud object storage (e.g. buckets, S3, Blob, GS).
    - SQL (e.g. PostgreSQL, MySQL)
    - SQL - embedded (e.g. SQLite)
    - NoSQL - columnar store (e.g. Parquet, Arrow, HDFS, BigQuery).
    - NoSQL - document store (e.g. MongoDB, Elasticsearch, CouchDB).
    - Graph database (Neo4j, TigerGraph).
    - Time Series (e.g. InfluxDB).
    - Pub sub (e.g. Kafka, Druid).
    - Key value (e.g. Redis, MemcacheDB).
    - Google Sheets.
    - ❗Industry or field specific APIs.
    - Streaming.
    - Other (please specify)

9. What data formats are you mostly working with? (pick up to 3)

>     - Tabular (csv, spreadsheet, database tables, parquet).

Reviewer 4: Wondering why some questions include `e.g.` when listing examples, but others do not 🤔 ?

Reviewer 2: ☑️ changes integrated. Used 'e.g.' everywhere; it reflects syntax/ cody tidiness.

11. What type of analysis are you running? (up to 4)
    - ❗I am not performing ML/statistical tasks.
    - Regression (predict a numeric output).
    - Classification (predict a categorical output).
    - Generative/ auto-encode (create new data based on existing data).
    - Reinforcement learning (actions that maximize a reward)
    - Dimensionality reduction (PCA, K-Nearest Neighbors)
    - Feature engineering (importance, extraction, selection, permutation).
    - Natural language processing.
    - Graph data science.
    - Outlier detection.
    - Other (please specify)
    
    
### Visualization

12. What tools does your team use to create dashboards tools? (pick up to 3)
    - ❗My team doesn’t create dashboard tools.
    - Write our own in HTML & JS.
    - R Shiny.
    - Kibana.
    - Dash-Plotly.
    - Voila.
    - Tableau.
    - Looker.
    - Klipfolio.
    - Google Data Studio.
    - Spotfire.
    - Grafana
    - Other (please specify)

13. Do you experience these problems with visualization in Jupyter?
    - No built-in UI for creating charts.
    - Can't publish my charts as web-based dashboards.
    - Poor/ buggy support for my plotting tool.

>     - Displaying highly dimensional data (array of array of arrays, too many rows/ columns to fit on screen).

Reviewer 4: All the other questions seem to be written from a "negative"  point of view, except this one

Reviewer 2: ☑️ changes integrated to each problem-style question

    - Lacking templating support (Jinja2)

    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

### Scale

14. How do you scale and schedule your workloads? (up to 4)
    - They run just fine on my local machine 😃.
    - ❓I need to scale, but don't know how.
    - On-premise HPC/ data center.

>     - Cloud server (e.g. EC2)

Reviewer 4: EC2 -> Amazon EC2 ?

Reviewer 2: ☑️ changes integrated; prefaced both occurrences of 'EC2' with 'AWS'

    - Cloud service for ML/ AI (e.g. AWS SageMaker)
    - Spark cluster (e.g. similar Hadoop, Dask).
    - K8s cluster (e.g. Mesos, Swarm, Slurm).
    - Jupyter Enterprise Gateway.
    - Jupyter BinderHub.
    - Quantum (e.g. D-Wave).
    - Horovod.
    - Kubeflow.
    - Elyra.
    - Workflow - Snakemake.
    - Workflow - Papermill
    - Workflow - Nextflow, WDL, CWL.
    - Workflow - Airflow.
    - Workflow - cloud service (e.g. AWS Batch)
    - Cloud query service (AWS Presto or Athena)
    - Other (please specify)

15. Do you experience these problems with scale in Jupyter?
    - Figuring out how to schedule batch execution of notebook-based jobs.
    - Don’t have the budget for more scalable environment/ cloud services.
    - Haven’t divided longer notebooks into multiple, modular notebooks.
    - Not persisting the outputs of a notebook.
    - ML training jobs take too long.
    - Can't call code/ modules from other notebooks.
    - Managing Spark dependencies (Java) and connections.

    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

### Collaboration

> 16. With how many other people are you collaborating?

Reviewer 4: collaborating on 🙃 ?

Reviewer 3: On notebooks

Reviewer 2: agree, let's make it more explicit about "notebooks"

Reviewer 2: ☑️ changes integrated: "When it comes to working on notebooks in a team setting, with how many other people are you collaborating?"

    - Don't know what dependencies (versions of language, packages, extensions) a notebook uses.
    - Don't know/ have the data a notebook is supposed to use.
    - Poor support for our version control (git) system.
    - Publishing my notebook to a shared location.
    - Not being able to comment on notebooks.
    - More robust non-git version history of a notebook.

    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

20. Do you have challenges with the notebook UI?
    - Poor autocompletion [LSP, show methods/ attributes].
    - No modes for editing other Jupyter documents (MyST, Jupyter Book).
    - Can't see hidden (.) files in file browser.
    - Can't collapse sections of a notebook hierarchically.
    - No progress bar for running long notebooks.
    - Don't know which cell failed in long notebook.
    - No Extension marketplace (5 star ratings, categories).
    - No global search.
    
    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

> 

Reviewer 4: I know we are on voting stage but I was wondering about the mobile story. Would it make sense to add a question on what are the expectations of Jupyter users regarding mobile, or something like that in terms of the different areas?

Reviewer: Mobile use (desired or attempted) would be very interesting to find out. We've long said that we were not focusing on mobile with jlab, however some people have made things work to some degree. If there is a huge community interest in having jlab work on mobile, that would be very interesting.

Reviewer 2: Hmm. One approach would be to add mobile devices to the "How do you access Jupyter" question with a callout for 'comments welcome about the experience.'

Reviewer 2: ![image](https://user-images.githubusercontent.com/16389820/98454852-3588ae80-2137-11eb-82e6-0ece207e8b1c.png)
☑️ changes integrated

# JupyterLab

 - Even if you don't use Jupyter, you can still take this survey. Just indicate this fact in the first question 
   and carry on as best as you can.
 - Thank you. Your participation guides Jupyter's roadmap toward your real-life use cases by quantifiably 
   helping us prioritize the functionality that is important to our userbase. 
 - So that you know what to expect, it's comprised of 20 questions spread across the sections below. 
   As a fair heads up, Question 7  is the biggest one, but it provides critical information.
    - Usage patterns.
    - Data
    - Visualization.
    - Scale.
    - Collaboration.
 - The aggregate survey data itself will be openly shared with the Jupyter community when polling closes in mid-December.

>    If you opt to provide your email address for a user interview, it not be used for Jupyter's promotional purposes and 

Reviewer 5: Missing word.
```suggestion
   If you opt to provide your email address for a user interview, it will not be used for Jupyter's promotional purposes and 
```

Reviewer 2: Thank you. Added 'will'

Reviewer 6: @LayneSadler Did you push these changes for missing words? I don't see them on the PR

Reviewer 2: @marthacryan No, I did not.

    - 1-2 years.
    - 6-12 months.
    - Less than 6 months (welcome =]).
    - I don't use Jupyter.

3. What languages do you use in Jupyter? (pick up to 4)
    - C (and derivatives)
    - Go
    - Groovy
    - Java
    - JavaScript
    - Julia
    - NodeJS
    - Perl
    - PHP
    - Python
    - R
    - Ruby
    - Rust
    - Scala
    - Spark SQL
    - SQL
    - TypeScript
    - ❗I wrap/ use bindings for other languages.
    - ❗My preferred language is not supported in Jupyter.
    - Other (please specify)

4. What are your primary job roles when you are using Jupyter? (pick up to 2)
    - Backend engineer.
    - Business analyst.
    - Data engineer.
    - Data scientist.
    - Database Admin (DBA).
    - DevOps.
    - Financial modeler/ analyst.
    - Front end/ web development.
    - Infrastructure engineer/ cloud architect.
    - Scientist/ researcher.
    - Student.
    - Sysadmin.
    - Teacher/ lecturer.
    - Tutor/ teaching assistant.
    - Other (please specify)

5. What are your go-to tools for performing data science, scientific computing, 
and machine learning on your laptop/ desktop (non-cloud) for data science? (pick up to 3)
    - Atom.
    - Emacs.

>     - iPython.

Author: this should be `IPython`

    - Jupyter Notebook - Classic.
    - JupyterLab.
    - nteract.
    - PyCharm.
    - RStudio.
    - Spyder.
    - Sublime Text.
    - Vim.
    - VS Code.
    - Zeppelin.
    - Other (please specify).

> 6. How do you run and/ or access Jupyter? (pick up to 4)

Author: I believe this is one of those questions that the alphabetical sort order might not be best, as a great percentage of users might be using Jupyter in their local env of some sort which would be the last answer or towards the end? Having said that, I don't have strong feelings one way or another...

    - BinderHub / MyBinder.
    - Cloud server (e.g. AWS EC2).
    - Cloud service - AWS (e.g. EMR, SageMaker).
    - Cloud service - Azure (e.g. Notebooks, ML Studio).
    - Cloud service - Databricks.
    - Cloud service - Google (e.g. AI Platform, Dataproc).

>     - Cloud service - IBM (e.g. Watson).

Author: This should be `Watson Studio`

    - Images.
    - Tensors (e.g. manually handling PyTorch, Tensorflow inputs).
    - Nested (e.g. JSON, NoSQL document).
    - Hierarchical Data Format (e.g. HDF5 or similar).
    - Time series.
    - Text.
    - Audio.
    - Video.
    - 3D/ CAD.
    - Graph (e.g. nodes, edges).
    - Spatial/ geographic (e.g. coordinates, GIS).
    - Game/ reinforcement simulation.
    - ❗Industry-specific file formats.
    - Other (please specify)

10. Do you experience these **problems with data** in Jupyter? (rate from scale of 0-4)
    - No grid view for manipulating/ filtering dataframes and arrays.
    - Can’t see a list of my current variables.
    - Plaintext or environment variable management of database passwords/ keys/ secrets.
    - Lost data during failure or restart of kernel/ server.
    - Data is too big to fit into memory on my machine/ server.
    - Poor MVC/ ORM integrations (e.g. Django, Flask).
    - Managing database/ source connections and secrets.
    - Other (please specify)
    
    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

> 11. What type of analysis are you running? (up to 4)

Reviewer 5: Missing word.
```suggestion
11. What type of analysis are you running? (pick up to 4)
```

Reviewer 2: Thank you. I just went through and made sure each one included 'pick'

    - Dimensionality reduction (e.g. PCA, K-Nearest Neighbors)
    - Feature engineering (e.g. importance, extraction, selection, permutation).
    - Natural language processing (NLP).
    - Graph data science.
    - Outlier detection.
    - Other (please specify)
    
    
### Visualization

12. What tools does your team use to create dashboards tools? (pick up to 3)
    - Dash-Plotly.
    - Google Data Studio.
    - Grafana
    - Kibana.
    - Klipfolio.
    - Looker.
    - R Shiny.
    - Spotfire.
    - Tableau.
    - Voila.
    - ❗I don't create dashboards.
    - ❗I write my own in HTML & JS.
    - Other (please specify).

13. Do you experience these problems with visualization in Jupyter?
    - No built-in UI for creating charts.
    - Can't publish my charts as web-based dashboards.
    - Poor/ buggy support for my plotting tool.
    - Displaying highly dimensional data (e.g. array of array of arrays, too many rows/ columns to fit on screen).
    - Lacking templating support (Jinja2)

    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

### Scale

> 14. How do you scale and schedule your workloads? (up to 4)

Reviewer 5: Missing word.
```suggestion
14. How do you scale and schedule your workloads? (pick up to 4)
```

    - Cluster - Jupyter Enterprise Gateway.
    - Cluster - Kubernetes (or similar e.g. Mesos, Swarm, Slurm).
    - Cluster - Spark (or similar e.g. Hadoop, Dask).
    - Elyra.
    - Horovod.
    - Jupyter BinderHub.
    - Kubeflow.
    - Quantum (e.g. D-Wave).
    - Server - Cloud (e.g. AWS EC2).
    - Server - On premise HPC/ data center.
    - Service - Cloud Queries (e.g. AWS Presto, AWS Athena).
    - Service - ML/ AI (e.g. AWS SageMaker).

>     - Workflow - Airflow.

Author: This should be `Apache Airflow`

    - Workflow - Snakemake.
    - Workflow - Papermill.
    - Workflow - Nextflow, WDL, CWL.
    - Workflow - Apache Airflow.
    - Workflow - Cloud service (e.g. AWS Batch).
    - Service - Cloud Queries (e.g. AWS Presto, AWS Athena).
    - Other (please specify).

15. Do you experience these problems with scale in Jupyter?
    - Figuring out how to schedule batch execution of notebook-based jobs.
    - Don’t have the budget for more scalable environment/ cloud services.
    - Haven’t divided longer notebooks into multiple, modular notebooks.
    - Not persisting the outputs of a notebook.
    - Machine learning training jobs take too long.
    - Can't call code/ modules from other notebooks.
    - Difficulty managing Spark dependencies (Java).

    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

### Collaboration

16. When it comes to working on notebooks in a team setting, with how many other people are you collaborating?
    - 0
    - 10
    - 20
    - 30
    - 40
    - 50+

> 17. What is your reason for sharing a notebook with someone else? (up to 3)

Reviewer 5: Missing word.
```suggestion
17. What is your reason for sharing a notebook with someone else? (pick up to 3)
```

    - No built-in way to publish my notebook to a shared location.
    - Not being able to comment on notebooks.
    - No "track changes;" can't figure out what changed between notebook checkpoints/ versions.

    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

20. Do you have challenges with the notebook UI?
    - No progress bar for running long notebooks.
    - No marketplace for Extensions (e.g. 5 star ratings, browsable categories).
    - No global search.
    - Can't collapse sections of a notebook hierarchically.
    - Poor autocompletion (e.g. LSP, show methods/ attributes).
    - No modes for editing other Jupyter documents (MyST, Jupyter Book).
    - Can't see hidden (.) files in file browser.
    - Don't know which cell failed in long notebook.
    
    For each of the items above, specify:
    - Not a problem for me.	
    - Trivial.	
    - Minor.	
    - Major.	
    - Critical.	
    - N/A - skip, don't know.

### You did it - thank you!

21. Open feedback for problems/ pain points you didn't get to share.

> 22. Optional - email address for recontact. Do we have your permission to recontact you for a user interview? 
>     We're looking for qualitative feedback on JupyterLab, JupyterHub, and the JupyterLab developer experience.

Reviewer 5: Gathering emails requires trust, so I want to make sure it's clear enough to survey participants what we are asking them to do (user interview may not be a familiar term) and that leaving an email address is giving permission (right now the question asks for a yes/no but the answer space takes neither). I think something like this could work:

```suggestion
22. Optional - Are you interested in giving qualitative feedback on JupyterLab, JupyterHub, and the JupyterLab developer experience? If we have permission to contact you for follow-up questions, please leave your email address below.
```

Reviewer 2: Good thinking. I'll swap out the text.

### Scale

14. How do you scale and schedule your workloads? (up to 4)
    - 🖥️ They run just fine on my local machine.
    - ❓I need to scale, but don't know how.
    - Server - On premise HPC/ data center.
    - Server - Cloud (e.g. AWS EC2).
    - Service - ML/ AI (e.g. AWS SageMaker).
    - Cluster - Spark (or similar e.g. Hadoop, Dask).
    - Cluster - Kubernetes (or similar e.g. Mesos, Swarm, Slurm).
    - Cluster - Jupyter Enterprise Gateway.
    - Jupyter BinderHub.
    - Quantum (e.g. D-Wave).
    - Horovod.
    - Kubeflow.
    - Elyra.
    - Workflow - Snakemake.

>     - Workflow - Papermill.

Reviewer 3: It probably makes sense to add Dask as it's own workflow item as well.

Reviewer 2: @willingc I'll agree in splitting out Dask from Spark. Here, _workflow_ has a stricter definition of reproducible, parameterized, user-specified DAG-style tasks where it's less of a detailed script and more of a string of ordered tasks. So Prefect might make more sense. 
https://examples.dask.org/applications/prefect-etl.html

Reviewer 3: Totally up to you @LayneSadler where to put Dask. It would be good to include somewhere even if not under the DAG workflow items.
